domain/application/state/application.go

```diff
@@ -450,23 +449,17 @@ func (st *ApplicationState) insertUnit(
 		return errors.Trace(err)
 	}
 
-	// TODO - we are mirroring what's in mongo, hence the unit name is known.
-	// In future we'll need to use a sequence to get a new unit id.
-	if args.UnitName == nil {
-		return fmt.Errorf("must pass unit name when adding a new unit for application %q", appID)
-	}
-	unitName := *args.UnitName
-	createParams.Name = unitName
+	createParams.Name = args.UnitName
 
 	if err := tx.Query(ctx, createNodeStmt, createParams).Run(); err != nil {
-		return errors.Annotatef(err, "creating net node row for unit %q", unitName)
+		return errors.Annotatef(err, "creating net node row for unit %q", args.UnitName)
 	}
 	if err := tx.Query(ctx, createUnitStmt, createParams).Run(); err != nil {
-		return errors.Annotatef(err, "creating unit row for unit %q", unitName)
+		return errors.Annotatef(err, "creating unit row for unit %q", args.UnitName)
 	}
 	if args.CloudContainer != nil {
-		if err := st.upsertUnitCloudContainer(ctx, tx, unitName, nodeUUID.String(), args.CloudContainer); err != nil {
-			return errors.Annotatef(err, "creating cloud container row for unit %q", unitName)
+		if err := st.upsertUnitCloudContainer(ctx, tx, args.UnitName, nodeUUID.String(), args.CloudContainer); err != nil {
```

Comment:
This has been a topic of conversation - we had agreed that it's fine for optional attributes to be conditionally persisted in the state layer, since there's no transformation occurring.